### PR TITLE
mouse-actions: init at 0.4.4

### PR DESCRIPTION
**Pull request by yliceee**
## Description of changes
https://github.com/jersou/mouse-actions

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:1]: Use the new [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name) format so the package is automatically added to nixpkgs

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:9]:
```suggestion
    rev = "v${version}";
```

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:1]:
```suggestion
{ lib
, fetchFromGitHub
, rustPlatform
, pkg-config
, virtualgl
, libX11
, libXi
, libXtst
, libevdev
}:

```

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:15]:
```suggestion

  buildInputs = [
    libX11
    libXi
    libXtst
    libevdev
  ];

  nativeBuildInputs = [
    pkg-config
    virtualgl
  ];
```
Using separate lines makes it easier to see when new dependencies are added with git.

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:19]:
```suggestion
    description = "Execute commands from mouse events such as clicks/wheel on the side/corners of the screen, or drawing shapes";
```

File: pkgs/tools/inputmethods/mouse-actions/default.nix
Comment (donovanglover) [inline pkgs/tools/inputmethods/mouse-actions/default.nix:20]:
```suggestion
    homepage = "https://github.com/jersou/mouse-actions";
```

**Review by donovanglover — CHANGES_REQUESTED** [review]
Thanks for contributing to nixpkgs!

**Comment by rgri** [comment]
Alright thanks for the criticism! ^-^ I think I've resolved everything so far. However not too sure how I should organize my commits, is it ok for there to be so many? Specifically the last one which moves and renames `default.nix`.

**Comment by donovanglover** [comment]
1 for the maintainer and 1 for the package would be ideal :+1:

**Comment by rgri** [comment]
Alright, I think I've done it.

**Comment by donovanglover** [comment]
Result of `nixpkgs-review pr 265056` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mouse-actions</li>
  </ul>
</details>

I got it to run after changing `cargoHash` but it crashes for me. Don't know how we usually handle groups in nixpkgs.

```rust
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args] Grab Error: IoError(
        Os {
            code: 13,
            kind: PermissionDenied,
            message: "Permission denied",
        },
    )
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args] The user must be in the file group of /dev/input files, usually 'input' or 'plugdev' :
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args]   $ sudo usermod -a -G input $USER
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args]   $ sudo usermod -a -G plugdev $USER
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args]   $ sudo tee /etc/udev/rules.d/80-mouse-actions.rules <<<'KERNEL=="uinput", SUBSYSTEM=="misc", TAG+="uaccess", OPTIONS+="static_node=uinput"'
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args] Then restart to apply this user modifications.
[2023-11-02T19:16:00Z ERROR mouse_actions::process_args] See https://github.com/jersou/mouse-actions#requirement-
```

**Comment by rgri** [comment]
I wasn't sure how to do this either. The upstream doesn't give any solution besides just `usermod`.

File: pkgs/by-name/mo/mouse-actions/package.nix
Comment (donovanglover) [inline pkgs/by-name/mo/mouse-actions/package.nix:23]:
```suggestion
  cargoHash = "sha256-5SUVZlrXIPtlu9KBzucZDCp5t5t8Z4/Nfht2Pw5agVI=";
```

File: pkgs/by-name/mo/mouse-actions/package.nix
Comment (donovanglover) [inline pkgs/by-name/mo/mouse-actions/package.nix:42]:
```suggestion
    mainProgram = "mouse-actions";
  };
```
Was going to suggest `platforms = platforms.linux;` as well but upstream [claims](https://github.com/jersou/mouse-actions#platform-compatibility) it works on Mac / Windows

**Comment by donovanglover** [comment]
Unfortunately I can't approve since I'm unfamiliar with how to proceed from here. Maybe someone else can help out.

**Comment by jfvillablanca** [comment]
can you squash your commits from the review, ideally there should only be 2 commits

**Comment by FliegendeWurst** [comment]
> I got it to run after changing cargoHash but it crashes for me. Don't know how we usually handle groups in nixpkgs.

Usually a `programs.<name>.enable` or `services.<name>.enable` nixos option is added that configures the required permissions.

File: pkgs/by-name/mo/mouse-actions/package.nix
Comment (FliegendeWurst) [inline pkgs/by-name/mo/mouse-actions/package.nix:34]:
```suggestion
```
The package builds successfully without virtualgl.

**Comment by rgri** [comment]
> Usually a programs.<name>.enable or services.<name>.enable nixos option is added that configures the required permissions.
Alright, I'll look into this tomorrow morning. Would this mean that it technically doesn't work on distros which aren't NixOS, and I should update `meta.platforms`?

**Comment by FliegendeWurst** [comment]
> Alright, I'll look into this tomorrow morning. Would this mean that it technically doesn't work on distros which aren't NixOS

The package itself can be run on other distros, the user then just needs to set their groups and udev rules manually.

**Comment by rgri** [comment]
Still interested in this and have a pushed a draft of the module in the latest (to be squashed later) commit, but I don't know how to add the current user to a group.

**Comment by donovanglover** [comment]
Remember to add new modules to [`nixos/modules/module-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix)

**Comment by rgri** [comment]
> > I got it to run after changing cargoHash but it crashes for me. Don't know how we usually handle groups in nixpkgs.
> 
> Usually a `programs.<name>.enable` or `services.<name>.enable` nixos option is added that configures the required permissions.

@FliegendeWurst Do you have any examples of what you're talking about here? I can't find any modules in nixpkgs that add the current users to certain groups, which is what I interpreted this.

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
If this was adding a single service user or something, a module would make sense; however as is, any potential module would simply be duplicating `users.users.<names>.extraGroups`.

If you want, you can add a note in the `meta.longDescription` of the package mentioning which groups are needed, but I think a module is inappropriate.

**Comment by rgri** [comment]
That was my first instinct as well. I'll revert the module creation and add a note.

**Comment by donovanglover** [comment]
Maybe you can do something similar to [`glasgow.nix`](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/hardware/glasgow.nix)? Would also ensure that the `plugdev` group exists.

File: nixos/modules/programs/mouse-actions.nix
Comment (h7x4) [inline nixos/modules/programs/mouse-actions.nix:11]: I don't think this is possible, because the module cannot by itself know what user it needs to add to these groups. You could either add it as an option (something like `programs.mouse-actions.users = [ ]`) or just ensure `groups.plugdev` exists like some other modules do (e.g. hackrf, minipro, ...)

Comment (rgri) [reply nixos/modules/programs/mouse-actions.nix]: Right. I think a simple approach is to add a few udev rules and then include them like [glasgow](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/hardware/glasgow.nix). I've pushed my first attempt at that, it seems to be working.

**Comment by rgri** [comment]
Oki, unless anyone in this thread has objections, I've [reviewed](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions) the package and am going to add it to the [already-reviewed](https://discourse.nixos.org/t/prs-already-reviewed/2617/1204) thread.

**Review by FliegendeWurst — APPROVED** [review]
LGTM besides the commit message (all `Co-authored-by` lines should be at the very end).

**Comment by rgri** [comment]
Done! Will wait till later tonight to start pushing for a merge in case other contributors would like to chime in.

**Review by FliegendeWurst — APPROVED** [review]

**Comment by rgri** [comment]
Anything I can do at this point to help get this merged?

**Comment by ludat** [comment]
I also would like this merged, is there something missing?

File: pkgs/by-name/mo/mouse-actions/package.nix
Comment (h7x4) [inline pkgs/by-name/mo/mouse-actions/package.nix:29]: I'm not familiar with darwin, but this seems reasonable based on https://github.com/jersou/mouse-actions/blob/25b341c0bd8627e3a7bf0dacc7b01e7664369028/rdev/Cargo.toml#L25-L37

I think you'll have to add the darwin dependencies through an entry in `pkgs/top-level/all-packages.nix`

@wegank do you mind if I ping you for this?

```suggestion
  buildInputs = [ ]
    ++ (lib.optionals stdenv.isLinux [
      libX11
      libXi
      libXtst
      libevdev
    ])
    ++ (lib.optionals stdenv.isDarwin [
      Cocoa
      CoreGraphics
      CoreFoundation
    ]);
```

Comment (wegank) [reply pkgs/by-name/mo/mouse-actions/package.nix:29]: Adding `darwin.apple_sdk.frameworks.Cocoa` should suffice. However, I had to

- replace `Pid::from_raw(pid)` with `Pid::from_raw(pid as i32)` in `src/single_instance.rs` for the build to succeed, and,
- make the `assert_eq!` on floats looser in `test_points_to_angles`.

I wonder if upstream has really tested darwin support.

Comment (ludat) [reply pkgs/by-name/mo/mouse-actions/package.nix:29]: the author mentions that mac __should__ work but it looks like they don't have a mac machine to test so I wouldn't be so sure about it.

> ## Platform compatibility
> I only tested it on Linux + X11 and Linux + Wayland, but it should work on Mac, Windows.

Comment (rgri) [reply pkgs/by-name/mo/mouse-actions/package.nix:29]: Hi, sorry I didn't see this until now. I would definitely be willing to add these changes.

But, I am unsure what the protocol is- am I responsible for testing/patching this package on mac, just for nixpkgs, when the upstream seems uninsterested/unable to test for themselves?

Moreover, it seems that [macOS does not have udev rules](https://forums.developer.apple.com/forums/thread/79167), which according to upstream is a [necessary](https://github.com/jersou/mouse-actions?tab=readme-ov-file#requirement-) feature to use the program. 

Open to advice!

Comment (eclairevoyant) [reply pkgs/by-name/mo/mouse-actions/package.nix:29]: If upstream isn't supporting darwin then I'd say mark it as only supporting linux for now. To say it supports "all" platforms is a bit misleading

Comment (rgri) [reply pkgs/by-name/mo/mouse-actions/package.nix:29]: Alright, it seems that everyone in this thread agrees this package is effectively linux-only. I'll mark it as resolved.

**Review by wegank — APPROVED** [review]

File: nixos/modules/programs/mouse-actions.nix
Comment (eclairevoyant) [inline nixos/modules/programs/mouse-actions.nix:11]:
```suggestion
        mouse-actions udev rules. This is a prerequisite for using mouse-actions without being root.
```

File: nixos/modules/programs/mouse-actions.nix
Comment (eclairevoyant) [inline nixos/modules/programs/mouse-actions.nix:3]: This pattern is discouraged, see #208242. Best to use `inherit` in the below `let` block.

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

**Review by eclairevoyant — APPROVED** [review]
second commit msg can be cleaned up to just one line, otherwise lgtm

**Comment by rgri** [comment]
I removed the squashed messages, but am leaving the "Co-authored by [...]" lines. Is that OK?

File: nixos/modules/programs/mouse-actions.nix
Comment (donovanglover) [inline nixos/modules/programs/mouse-actions.nix:1]: Should be added to [`nixos/modules/module-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/module-list.nix)

File: pkgs/by-name/mo/mouse-actions/package.nix
Comment (donovanglover) [inline pkgs/by-name/mo/mouse-actions/package.nix:47]:
```suggestion
    platforms = platforms.linux;
```

**Comment by FliegendeWurst** [comment]
> I removed the squashed messages, but am leaving the "Co-authored by [...]" lines. Is that OK?

Yes, that is how it is supposed to be done.

**Review by donovanglover — APPROVED** [review]

**Pull request merged**
